diff -ur a/tclreadline.c b/tclreadline.c
--- a/tclreadline.c	2013-11-16 20:14:28.876272386 +0100
+++ b/tclreadline.c	2013-11-16 20:37:47.455986902 +0100
@@ -234,7 +234,7 @@
 
 	    rl_callback_handler_install(
 			       objc == 3 ? Tcl_GetStringFromObj(objv[2], 0)
-			       : "%", TclReadlineLineCompleteHandler);
+			       : "% ", TclReadlineLineCompleteHandler);
 
 	    Tcl_CreateFileHandler(0, TCL_READABLE,
 		TclReadlineReadHandler, (ClientData) NULL);
diff -ur a/tclreadlineSetup.tcl.in b/tclreadlineSetup.tcl.in
--- a/tclreadlineSetup.tcl.in	2013-11-16 20:36:31.546897755 +0100
+++ b/tclreadlineSetup.tcl.in	2013-11-16 20:37:21.882630726 +0100
@@ -165,7 +165,7 @@
 		    if [info exists env(HOME)] {
 			regsub $env(HOME) $pwd "~" pwd
 		    }
-		    return "$prompt_string \[$pwd\]"
+		    return "$prompt_string \[$pwd\] "
 		}
 	    }
 	    # puts body=[info body ::tclreadline::prompt1]
@@ -266,7 +266,7 @@
 	    if [info exists tcl_prompt2] {
 		set prompt2 $tcl_prompt2
 	    } else {
-		set prompt2 ">"
+		set prompt2 "> "
 	    }
 
 	    if {[catch {
@@ -274,7 +274,7 @@
 		    set LINE [::tclreadline::readline read \
 		    [::tclreadline::prompt1]]
 		} else {
-		    set LINE [::tclreadline::readline read %]
+		    set LINE [::tclreadline::readline read "% "]
 		}
 		while {![::tclreadline::readline complete $LINE]} {
 		    append LINE "\n"
