diff --git a/src/streams/openssl.c b/src/streams/openssl.c
index 9cbb274..8b4dd68 100644
--- a/src/streams/openssl.c
+++ b/src/streams/openssl.c
@@ -150,7 +150,7 @@ int git_openssl_stream_global_init(void)
 	return 0;
 }
 
-#if defined(GIT_THREADS)
+#if defined(GIT_THREADS) && OPENSSL_VERSION_NUMBER < 0x10100000L
 static void threadid_cb(CRYPTO_THREADID *threadid)
 {
     CRYPTO_THREADID_set_numeric(threadid, git_thread_currentid());
