diff --git a/man/color.map.5 b/man/color.map.5
index 5543628..e371e2d 100644
--- a/man/color.map.5
+++ b/man/color.map.5
@@ -76,6 +76,18 @@ Defines color used for world packages not planned to be merged.
 Defines color used for packages planned to be uninstalled in order
 to resolve conflicts.
 .TP
+\fBUSE_ENABLED\fR = \fI"red"\fR
+Defines color used for a package's enabled USE flags.
+.TP
+\fBUSE_DISABLED\fR = \fI"blue"\fR
+Defines color used for a package's disabled USE flags.
+.TP
+\fBUSE_NEW\fR = \fI"yellow"\fR
+Defines color used for a package's newly available USE flags.
+.TP
+\fBUSE_TOGGLED\fR = \fI"green"\fR
+Defines color used for a reinstalled package's toggled USE flags.
+.TP
 \fBPROMPT_CHOICE_DEFAULT\fR = \fI"green"\fR
 Defines color used for the default choice at a prompt.
 .TP
diff --git a/pym/_emerge/resolver/output_helpers.py b/pym/_emerge/resolver/output_helpers.py
index 70f7bc0..ef9d06f 100644
--- a/pym/_emerge/resolver/output_helpers.py
+++ b/pym/_emerge/resolver/output_helpers.py
@@ -253,14 +253,14 @@ def _create_use_string(conf, name, cur_iuse, iuse_forced, cur_use,
 			isEnabled = True
 			if is_new or flag in old_use and \
 				(conf.all_flags or reinst_flag):
-				flag_str = red(flag)
+				flag_str = colorize("USE_ENABLED", flag)
 			elif flag not in old_iuse:
-				flag_str = yellow(flag) + "%*"
+				flag_str = colorize("USE_NEW", flag) + "%*"
 			elif flag not in old_use:
-				flag_str = green(flag) + "*"
+				flag_str = colorize("USE_TOGGLED", flag) + "*"
 		elif flag in removed_iuse:
 			if conf.all_flags or reinst_flag:
-				flag_str = yellow("-" + flag) + "%"
+				flag_str = colorize("USE_NEW", "-" + flag) + "%"
 				if flag in old_use:
 					flag_str += "*"
 				flag_str = "(" + flag_str + ")"
@@ -270,13 +270,13 @@ def _create_use_string(conf, name, cur_iuse, iuse_forced, cur_use,
 			if is_new or flag in old_iuse and \
 				flag not in old_use and \
 				(conf.all_flags or reinst_flag):
-				flag_str = blue("-" + flag)
+				flag_str = colorize("USE_DISABLED", "-" + flag)
 			elif flag not in old_iuse:
-				flag_str = yellow("-" + flag)
+				flag_str = colorize("USE_NEW", "-" + flag)
 				if flag not in iuse_forced:
 					flag_str += "%"
 			elif flag in old_use:
-				flag_str = green("-" + flag) + "*"
+				flag_str = colorize("USE_TOGGLED", "-" + flag) + "*"
 		if flag_str:
 			if flag in feature_flags:
 				flag_str = "{" + flag_str + "}"
diff --git a/pym/portage/output.py b/pym/portage/output.py
index 7846627..be36357 100644
--- a/pym/portage/output.py
+++ b/pym/portage/output.py
@@ -144,6 +144,10 @@ _styles["PKG_UNINSTALL"]           = ( "red", )
 _styles["PKG_NOMERGE"]             = ( "darkblue", )
 _styles["PKG_NOMERGE_SYSTEM"]      = ( "darkblue", )
 _styles["PKG_NOMERGE_WORLD"]       = ( "blue", )
+_styles["USE_ENABLED"]             = ( "red", )
+_styles["USE_DISABLED"]            = ( "blue", )
+_styles["USE_NEW"]                 = ( "yellow", )
+_styles["USE_TOGGLED"]             = ( "green", )
 _styles["PROMPT_CHOICE_DEFAULT"]   = ( "green", )
 _styles["PROMPT_CHOICE_OTHER"]     = ( "red", )
 
