diff --git a/man/color.map.5 b/man/color.map.5
index 19242ac..77f6cff 100644
--- a/man/color.map.5
+++ b/man/color.map.5
@@ -112,6 +112,9 @@ Defines color used for a package's newly available USE flags.
 \fBUSE_TOGGLED\fR = \fI"green"\fR
 Defines color used for a reinstalled package's toggled USE flags.
 .TP
+\fBREPOSITORY\fR = \fI"teal"\fR
+Defines color used for the list of repositories.
+.TP
 \fBPROMPT_CHOICE_DEFAULT\fR = \fI"green"\fR
 Defines color used for the default choice at a prompt.
 .TP
diff --git a/pym/_emerge/resolver/output_helpers.py b/pym/_emerge/resolver/output_helpers.py
index 7248f03..b97ca80 100644
--- a/pym/_emerge/resolver/output_helpers.py
+++ b/pym/_emerge/resolver/output_helpers.py
@@ -18,8 +18,7 @@ from portage import os
 from portage import _encodings, _unicode_encode
 from portage._sets.base import InternalPackageSet
 from portage.localization import localized_size
-from portage.output import (bold, colorize, create_color_func,
-	teal)
+from portage.output import (bold, colorize, create_color_func)
 bad = create_color_func("BAD")
 from portage.util import writemsg
 from portage.util.SlotObject import SlotObject
@@ -75,9 +74,9 @@ class _RepoDisplay(object):
 			show_repo_paths[repo_index] = repo_path
 		if show_repo_paths:
 			for index, repo_path in enumerate(show_repo_paths):
-				output.append(" "+teal("["+str(index)+"]")+" %s\n" % repo_path)
+				output.append(" "+colorize("REPOSITORY", "["+str(index)+"]")+" %s\n" % repo_path)
 		if unknown_repo:
-			output.append(" "+teal("[?]") + \
+			output.append(" "+colorize("REPOSITORY", "[?]") + \
 				" indicates that the source repository could not be determined\n")
 		return "".join(output)
 
diff --git a/pym/portage/output.py b/pym/portage/output.py
index 93494e5..a25b29c 100644
--- a/pym/portage/output.py
+++ b/pym/portage/output.py
@@ -156,6 +156,7 @@ _styles["USE_ENABLED"]             = ( "red", )
 _styles["USE_DISABLED"]            = ( "blue", )
 _styles["USE_NEW"]                 = ( "yellow", )
 _styles["USE_TOGGLED"]             = ( "green", )
+_styles["REPOSITORY"]              = ( "teal", )
 _styles["PROMPT_CHOICE_DEFAULT"]   = ( "green", )
 _styles["PROMPT_CHOICE_OTHER"]     = ( "red", )
 
