diff --git a/modules/videoio/src/cap_ffmpeg_impl.hpp b/modules/videoio/src/cap_ffmpeg_impl.hpp
index 8ca304a..19f6253 100644
--- a/modules/videoio/src/cap_ffmpeg_impl.hpp
+++ b/modules/videoio/src/cap_ffmpeg_impl.hpp
@@ -172,6 +172,14 @@ extern "C" {
 #endif
 #endif
 
+#if LIBAVFORMAT_VERSION_MAJOR < 58
+#  define AV_CODEC_FLAG_GLOBAL_HEADER CODEC_FLAG_GLOBAL_HEADER
+#endif
+
+#if LIBAVFORMAT_VERSION_MAJOR < 58
+#  define AV_PKT_FLAG_KEY PKT_FLAG_KEY
+#endif
+
 #if USE_AV_INTERRUPT_CALLBACK
 #define LIBAVFORMAT_INTERRUPT_OPEN_TIMEOUT_MS 30000
 #define LIBAVFORMAT_INTERRUPT_READ_TIMEOUT_MS 30000
@@ -1655,7 +1663,7 @@ static int icv_av_write_frame_FFMPEG( AVFormatContext * oc, AVStream * video_st,
             pkt.pts = c->coded_frame->pts;
 #endif
             if(c->coded_frame->key_frame)
-                pkt.flags |= PKT_FLAG_KEY;
+                pkt.flags |= AV_PKT_FLAG_KEY;
             pkt.stream_index= video_st->index;
             pkt.data= outbuf;
             pkt.size= out_size;
@@ -2504,7 +2512,7 @@ void OutputMediaStream_FFMPEG::write(unsigned char* data, int size, int keyFrame
         av_init_packet(&pkt);
 
         if (keyFrame)
-            pkt.flags |= PKT_FLAG_KEY;
+            pkt.flags |= AV_PKT_FLAG_KEY;
 
         pkt.stream_index = video_st_->index;
         pkt.data = data;
