diff --git a/configure b/configure
index 98d0c68f..499f6c81 100755
--- a/configure
+++ b/configure
@@ -855,7 +855,11 @@ check_ssl()
 extern "C" int main(void)
 {
   printf("found version %s", OPENSSL_VERSION_TEXT);
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
   SSL_library_init();
+#else
+  OPENSSL_init_ssl(0, NULL);
+#endif
 #if OPENSSL_VERSION_NUMBER >= 0x00908000
   printf(", OK.\n");
   return 0;
diff --git a/src/VBox/RDP/client-1.8.3/ssl.h b/src/VBox/RDP/client-1.8.3/ssl.h
index 1a034428..13b4d806 100644
--- a/src/VBox/RDP/client-1.8.3/ssl.h
+++ b/src/VBox/RDP/client-1.8.3/ssl.h
@@ -36,6 +36,7 @@
 #include <openssl/bn.h>
 #include <openssl/x509v3.h>
 #include <openssl/hmac.h>
+#include <openssl/rsa.h>
 
 #if defined(OPENSSL_VERSION_NUMBER) && (OPENSSL_VERSION_NUMBER >= 0x0090800f)
 #define D2I_X509_CONST const
diff --git a/src/VBox/RDP/client-1.8.3/tcp.c b/src/VBox/RDP/client-1.8.3/tcp.c
index 326b6c46..7cbf631b 100644
--- a/src/VBox/RDP/client-1.8.3/tcp.c
+++ b/src/VBox/RDP/client-1.8.3/tcp.c
@@ -311,8 +311,12 @@ tcp_tls_connect(void)
 
 	if (!g_ssl_initialized)
 	{
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
 		SSL_load_error_strings();
 		SSL_library_init();
+#else
+		OPENSSL_init_ssl(OPENSSL_INIT_LOAD_SSL_STRINGS, NULL);
+#endif
 		g_ssl_initialized = True;
 	}
 
@@ -325,7 +329,11 @@ tcp_tls_connect(void)
 #  pragma GCC diagnostic ignored "-Wdeprecated-declarations"
 # endif
 #endif
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
 		g_ssl_ctx = SSL_CTX_new(TLSv1_client_method());
+#else
+		g_ssl_ctx = SSL_CTX_new(TLS_client_method());
+#endif
 #ifdef __GNUC__
 # if (__GNUC__ << 8) + __GNUC_MINOR__ >= 0x406
 #  pragma GCC diagnostic pop
diff --git a/src/VBox/Runtime/VBox/VBoxRTDeps.cpp b/src/VBox/Runtime/VBox/VBoxRTDeps.cpp
index cfe31674..b0ceb97e 100644
--- a/src/VBox/Runtime/VBox/VBoxRTDeps.cpp
+++ b/src/VBox/Runtime/VBox/VBoxRTDeps.cpp
@@ -49,6 +49,7 @@
 #include <openssl/rsa.h>
 #include <openssl/ssl.h>
 #include <openssl/rand.h>
+#include <openssl/dh.h>
 
 
 /*********************************************************************************************************************************
diff --git a/src/VBox/Runtime/common/crypto/iprt-openssl.cpp b/src/VBox/Runtime/common/crypto/iprt-openssl.cpp
index b7719ea9..7a5de77d 100644
--- a/src/VBox/Runtime/common/crypto/iprt-openssl.cpp
+++ b/src/VBox/Runtime/common/crypto/iprt-openssl.cpp
@@ -44,9 +44,13 @@ DECLHIDDEN(void) rtCrOpenSslInit(void)
     static bool s_fOssInitalized;
     if (!s_fOssInitalized)
     {
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
         OpenSSL_add_all_algorithms();
         ERR_load_ERR_strings();
         ERR_load_crypto_strings();
+#else
+        OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CRYPTO_STRINGS | OPENSSL_INIT_ADD_ALL_CIPHERS | OPENSSL_INIT_ADD_ALL_DIGESTS, NULL);
+#endif
 
         s_fOssInitalized = true;
     }
