diff --git a/src/openssl.c b/src/openssl.c
index 4f63ef9..a3b490b 100644
--- a/src/openssl.c
+++ b/src/openssl.c
@@ -1068,7 +1068,11 @@ _libssh2_pub_priv_keyfile(LIBSSH2_SESSION *session,
          * ($#&%#$(%$#( Someone buy me an OpenSSL manual and I'll read up on
          * it.
          */
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
         OpenSSL_add_all_ciphers();
+#else
+        OPENSSL_init_crypto(OPENSSL_INIT_ADD_ALL_CIPHERS, NULL);
+#endif
     }
     BIO_reset(bp);
     pk = PEM_read_bio_PrivateKey(bp, NULL, NULL, (void*)passphrase);
@@ -1144,7 +1148,11 @@ _libssh2_pub_priv_keyfilememory(LIBSSH2_SESSION *session,
          * ($#&%#$(%$#( Someone buy me an OpenSSL manual and I'll read up on
          * it.
          */
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
         OpenSSL_add_all_ciphers();
+#else
+        OPENSSL_init_crypto(OPENSSL_INIT_ADD_ALL_CIPHERS, NULL);
+#endif
     }
     BIO_reset(bp);
     pk = PEM_read_bio_PrivateKey(bp, NULL, NULL, (void*)passphrase);
diff --git a/src/openssl.h b/src/openssl.h
index 3ca71fa..12419ca 100644
--- a/src/openssl.h
+++ b/src/openssl.h
@@ -226,10 +226,17 @@ int _libssh2_md5_init(libssh2_md5_ctx *ctx);
 #define libssh2_hmac_cleanup(ctx) HMAC_cleanup(ctx)
 #endif
 
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
 #define libssh2_crypto_init() \
   OpenSSL_add_all_algorithms(); \
   ENGINE_load_builtin_engines(); \
   ENGINE_register_all_complete()
+#else
+#define libssh2_crypto_init() \
+  OPENSSL_init_crypto(OPENSSL_INIT_ADD_ALL_CIPHERS | OPENSSL_INIT_ADD_ALL_DIGESTS, NULL); \
+  ENGINE_load_builtin_engines(); \
+  ENGINE_register_all_complete()
+#endif
 
 #define libssh2_crypto_exit()
 
