diff --git a/src/node_crypto.cc b/src/node_crypto.cc
index 111227e..c7f1c56 100644
--- a/src/node_crypto.cc
+++ b/src/node_crypto.cc
@@ -42,6 +42,10 @@
 #include <math.h>
 #include <string.h>
 
+#include <openssl/bn.h>
+#include <openssl/dh.h>
+#include <openssl/rsa.h>
+
 #include <algorithm>
 #include <memory>
 #include <vector>
@@ -1678,7 +1682,11 @@ static Local<Object> X509ToObject(Environment* env, X509* cert) {
     rsa = nullptr;
   }
 
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
   ASN1_TIME_print(bio, X509_get_notBefore(cert));
+#else
+  ASN1_TIME_print(bio, X509_get0_notBefore(cert));
+#endif
   BIO_get_mem_ptr(bio, &mem);
   info->Set(context, env->valid_from_string(),
             String::NewFromUtf8(env->isolate(), mem->data,
@@ -1686,7 +1694,11 @@ static Local<Object> X509ToObject(Environment* env, X509* cert) {
                                 mem->length)).FromJust();
   USE(BIO_reset(bio));
 
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
   ASN1_TIME_print(bio, X509_get_notAfter(cert));
+#else
+  ASN1_TIME_print(bio, X509_get0_notAfter(cert));
+#endif
   BIO_get_mem_ptr(bio, &mem);
   info->Set(context, env->valid_to_string(),
             String::NewFromUtf8(env->isolate(), mem->data,
@@ -5356,8 +5368,13 @@ void TimingSafeEqual(const FunctionCallbackInfo<Value>& args) {
 }
 
 void InitCryptoOnce() {
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
   SSL_load_error_strings();
   OPENSSL_no_config();
+#else
+  OPENSSL_init_ssl(OPENSSL_INIT_LOAD_CRYPTO_STRINGS | OPENSSL_INIT_LOAD_SSL_STRINGS
+                   | OPENSSL_INIT_NO_LOAD_CONFIG, NULL);
+#endif
 
   // --openssl-config=...
   if (!openssl_config.empty()) {
@@ -5379,8 +5396,12 @@ void InitCryptoOnce() {
     }
   }
 
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
   SSL_library_init();
   OpenSSL_add_all_algorithms();
+#else
+  OPENSSL_init_ssl(OPENSSL_INIT_ADD_ALL_CIPHERS | OPENSSL_INIT_ADD_ALL_DIGESTS, NULL);
+#endif
 
 #ifdef NODE_FIPS_MODE
   /* Override FIPS settings in cnf file, if needed. */
diff --git a/src/node_crypto.h b/src/node_crypto.h
index 2f7c904..5bf9448 100644
--- a/src/node_crypto.h
+++ b/src/node_crypto.h
@@ -50,6 +50,7 @@
 #include <openssl/hmac.h>
 #include <openssl/rand.h>
 #include <openssl/pkcs12.h>
+#include <openssl/dh.h>
 
 #if !defined(OPENSSL_NO_TLSEXT) && defined(SSL_CTX_set_tlsext_status_cb)
 # define NODE__HAVE_TLSEXT_STATUS_CB
