diff --git a/lib/compat/openssl_support.c b/lib/compat/openssl_support.c
index 4178ab0..9a6488a 100644
--- a/lib/compat/openssl_support.c
+++ b/lib/compat/openssl_support.c
@@ -122,7 +122,7 @@ _init_threadid_callback(void)
   CRYPTO_set_id_callback(_ssl_thread_id);
 }
 
-#else
+#elif OPENSSL_VERSION_NUMBER < 0x10100000L
 
 static void
 _ssl_thread_id2(CRYPTO_THREADID *id)
@@ -136,6 +136,13 @@ _init_threadid_callback(void)
   CRYPTO_THREADID_set_callback(_ssl_thread_id2);
 }
 
+#else
+
+static void
+_init_threadid_callback(void)
+{
+}
+
 #endif
 
 void
diff --git a/lib/tlscontext.c b/lib/tlscontext.c
index c45de28..c4b56ca 100644
--- a/lib/tlscontext.c
+++ b/lib/tlscontext.c
@@ -477,7 +477,7 @@ _load_dh_fallback(void)
   BIGNUM *g = NULL;
   BN_dec2bn(&g, "2");
 
-  if (!DH_set0_pqg(dh, get_rfc3526_prime_2048(NULL), NULL, g))
+  if (!DH_set0_pqg(dh, BN_get_rfc3526_prime_2048(NULL), NULL, g))
     {
       BN_free(g);
       DH_free(dh);
