diff --git a/imap-send.c b/imap-send.c
index 36c7c1b..17419aa 100644
--- a/imap-send.c
+++ b/imap-send.c
@@ -284,6 +284,7 @@ static int ssl_socket_connect(struct imap_socket *sock, int use_tls_only, int ve
 	int ret;
 	X509 *cert;
 
+#if (OPENSSL_VERSION_NUMBER < 0x10100000L)
 	SSL_library_init();
 	SSL_load_error_strings();
 
@@ -292,6 +293,13 @@ static int ssl_socket_connect(struct imap_socket *sock, int use_tls_only, int ve
 		ssl_socket_perror("SSLv23_method");
 		return -1;
 	}
+#else
+	meth = TLS_method();
+	if (!meth) {
+		ssl_socket_perror("TLS_method");
+		return -1;
+	}
+#endif
 
 	ctx = SSL_CTX_new(meth);
 	if (!ctx) {
@@ -300,7 +308,11 @@ static int ssl_socket_connect(struct imap_socket *sock, int use_tls_only, int ve
 	}
 
 	if (use_tls_only)
+#if (OPENSSL_VERSION_NUMBER < 0x10100000L)
 		SSL_CTX_set_options(ctx, SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3);
+#else
+		SSL_CTX_set_min_proto_version(ctx, TLS1_VERSION);
+#endif
 
 	if (verify)
 		SSL_CTX_set_verify(ctx, SSL_VERIFY_PEER, NULL);
