diff --git a/configure.ac b/configure.ac
index e50c3d7..12c772a 100644
--- a/configure.ac
+++ b/configure.ac
@@ -2547,6 +2547,7 @@ if test "x$openssl" = "xyes" ; then
 			)
 		]
 	)
+	AC_CHECK_FUNCS([SSLeay OpenSSL_version_num SSLeay_add_all_algorithms OPENSSL_init_crypto])
 
 	# Determine OpenSSL header version
 	AC_MSG_CHECKING([OpenSSL header version])
@@ -2602,8 +2603,15 @@ if test "x$openssl" = "xyes" ; then
 		if(fd == NULL)
 			exit(1);
 
+		#if defined(HAVE_SSLEAY)
 		if ((rc = fprintf(fd, "%08lx (%s)\n", (unsigned long)SSLeay(),
 		    SSLeay_version(SSLEAY_VERSION))) < 0)
+		#elif defined(HAVE_OPENSSL_VERSION_NUM)
+		if ((rc = fprintf(fd, "%08lx (%s)\n", (unsigned long)OpenSSL_version_num(),
+		    OpenSSL_version(OPENSSL_VERSION))) < 0)
+		#else
+		#error
+		#endif
 			exit(1);
 
 		exit(0);
@@ -2645,7 +2653,13 @@ if test "x$openssl" = "xyes" ; then
 	#include <openssl/opensslv.h>
 	#include <openssl/crypto.h>
 		]], [[
+		#if defined(HAVE_SSLEAY)
 		exit(SSLeay() == OPENSSL_VERSION_NUMBER ? 0 : 1);
+		#elif defined(HAVE_OPENSSL_VERSION_NUM)
+		exit(OpenSSL_version_num() == OPENSSL_VERSION_NUMBER ? 0 : 1);
+		#else
+		#error
+		#endif
 		]])],
 		[
 			AC_MSG_RESULT([yes])
@@ -2673,7 +2687,15 @@ if test "x$openssl" = "xyes" ; then
 	AC_MSG_CHECKING([if programs using OpenSSL functions will link])
 	AC_LINK_IFELSE(
 		[AC_LANG_PROGRAM([[ #include <openssl/evp.h> ]],
-		[[ SSLeay_add_all_algorithms(); ]])],
+		[[
+		#if defined(HAVE_SSLEAY_ADD_ALL_ALGORITHMS)
+		SSLeay_add_all_algorithms();
+		#elif defined(HAVE_OPENSSL_INIT_CRYPTO)
+		OPENSSL_init_crypto(0, NULL);
+		#else
+		#error
+		#endif
+		]])],
 		[
 			AC_MSG_RESULT([yes])
 		],
@@ -2684,7 +2706,15 @@ if test "x$openssl" = "xyes" ; then
 			AC_MSG_CHECKING([if programs using OpenSSL need -ldl])
 			AC_LINK_IFELSE(
 				[AC_LANG_PROGRAM([[ #include <openssl/evp.h> ]],
-				[[ SSLeay_add_all_algorithms(); ]])],
+				[[
+				#if defined(HAVE_SSLEAY_ADD_ALL_ALGORITHMS)
+				SSLeay_add_all_algorithms();
+				#elif defined(HAVE_OPENSSL_INIT_CRYPTO)
+				OPENSSL_init_crypto(0, NULL);
+				#else
+				#error
+				#endif
+				]])],
 				[
 					AC_MSG_RESULT([yes])
 				],
@@ -2697,6 +2727,8 @@ if test "x$openssl" = "xyes" ; then
 	)
 
 	AC_CHECK_FUNCS([ \
+		SSLeay_version \
+		OpenSSL_version \
 		BN_is_prime_ex \
 		DSA_generate_parameters_ex \
 		EVP_DigestInit_ex \
diff --git a/entropy.c b/entropy.c
index c178c00..d03a7aa 100644
--- a/entropy.c
+++ b/entropy.c
@@ -219,9 +219,17 @@ seed_rng(void)
 #ifndef OPENSSL_PRNG_ONLY
 	unsigned char buf[RANDOM_SEED_SIZE];
 #endif
+#if defined(HAVE_SSLEAY)
 	if (!ssh_compatible_openssl(OPENSSL_VERSION_NUMBER, SSLeay()))
 		fatal("OpenSSL version mismatch. Built against %lx, you "
 		    "have %lx", (u_long)OPENSSL_VERSION_NUMBER, SSLeay());
+#elif defined(HAVE_OPENSSL_VERSION_NUM)
+	if (!ssh_compatible_openssl(OPENSSL_VERSION_NUMBER, OpenSSL_version_num()))
+		fatal("OpenSSL version mismatch. Built against %lx, you "
+		    "have %lx", (u_long)OPENSSL_VERSION_NUMBER, OpenSSL_version_num());
+#else
+#error
+#endif
 
 #ifndef OPENSSL_PRNG_ONLY
 	if (RAND_status() == 1) {
diff --git a/ssh-keysign.c b/ssh-keysign.c
index 744ecb4..e921574 100644
--- a/ssh-keysign.c
+++ b/ssh-keysign.c
@@ -224,7 +224,9 @@ main(int argc, char **argv)
 		fatal("could not open any host key");
 
 #ifdef WITH_OPENSSL
+# if OPENSSL_VERSION_NUMBER < 0x10100000L
 	OpenSSL_add_all_algorithms();
+# endif
 	arc4random_buf(rnd, sizeof(rnd));
 	RAND_seed(rnd, sizeof(rnd));
 #endif
diff --git a/ssh.c b/ssh.c
index 0777c31..361d21d 100644
--- a/ssh.c
+++ b/ssh.c
@@ -816,7 +816,13 @@ main(int ac, char **av)
 			fprintf(stderr, "%s, %s\n",
 			    SSH_RELEASE,
 #ifdef WITH_OPENSSL
+# if defined(HAVE_SSLEAY_VERSION)
 			    SSLeay_version(SSLEAY_VERSION)
+# elif defined(HAVE_OPENSSL_VERSION)
+			    OpenSSL_version(OPENSSL_VERSION)
+# else
+#  error
+# endif
 #else
 			    "without OpenSSL"
 #endif
@@ -1036,7 +1042,7 @@ main(int ac, char **av)
 
 	host_arg = xstrdup(host);
 
-#ifdef WITH_OPENSSL
+#if defined(WITH_OPENSSL) && OPENSSL_VERSION_NUMBER < 0x10100000L
 	OpenSSL_add_all_algorithms();
 	ERR_load_crypto_strings();
 #endif
@@ -1085,7 +1091,13 @@ main(int ac, char **av)
 	if (debug_flag)
 		logit("%s, %s", SSH_RELEASE,
 #ifdef WITH_OPENSSL
+# if defined(HAVE_SSLEAY_VERSION)
 		    SSLeay_version(SSLEAY_VERSION)
+# elif defined(HAVE_OPENSSL_VERSION)
+		    OpenSSL_version(OPENSSL_VERSION)
+# else
+#  error
+# endif
 #else
 		    "without OpenSSL"
 #endif
diff --git a/ssh_api.c b/ssh_api.c
index c84b4e7..e03ddfa 100644
--- a/ssh_api.c
+++ b/ssh_api.c
@@ -79,9 +79,9 @@ ssh_init(struct ssh **sshp, int is_server, struct kex_params *kex_params)
 	int r;
 
 	if (!called) {
-#ifdef WITH_OPENSSL
+#if defined(WITH_OPENSSL) && OPENSSL_VERSION_NUMBER < 0x10100000L
 		OpenSSL_add_all_algorithms();
-#endif /* WITH_OPENSSL */
+#endif /* defined(WITH_OPENSSL) && OPENSSL_VERSION_NUMBER < 0x10100000L */
 		called = 1;
 	}
 
diff --git a/sshd.c b/sshd.c
index ba26287..afd21ed 100644
--- a/sshd.c
+++ b/sshd.c
@@ -917,7 +917,13 @@ usage(void)
 	fprintf(stderr, "%s, %s\n",
 	    SSH_RELEASE,
 #ifdef WITH_OPENSSL
-	    SSLeay_version(SSLEAY_VERSION)
+# if defined(HAVE_SSLEAY_VERSION)
+		    SSLeay_version(SSLEAY_VERSION)
+# elif defined(HAVE_OPENSSL_VERSION)
+		    OpenSSL_version(OPENSSL_VERSION)
+# else
+#  error
+# endif
 #else
 	    "without OpenSSL"
 #endif
@@ -1723,7 +1729,13 @@ main(int ac, char **av)
 
 	debug("sshd version %s, %s", SSH_VERSION,
 #ifdef WITH_OPENSSL
+# if defined(HAVE_SSLEAY_VERSION)
 	    SSLeay_version(SSLEAY_VERSION)
+# elif defined(HAVE_OPENSSL_VERSION)
+	    OpenSSL_version(OPENSSL_VERSION)
+# else
+#  error
+# endif
 #else
 	    "without OpenSSL"
 #endif
diff --git a/openbsd-compat/openssl-compat.c b/openbsd-compat/openssl-compat.c
index e9173a1..d717f9e 100644
--- a/openbsd-compat/openssl-compat.c
+++ b/openbsd-compat/openssl-compat.c
@@ -70,13 +70,13 @@ ssh_compatible_openssl(long headerver, long libver)
 void
 ssh_OpenSSL_add_all_algorithms(void)
 {
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
 	OpenSSL_add_all_algorithms();
 
 	/* Enable use of crypto hardware */
 	ENGINE_load_builtin_engines();
 	ENGINE_register_all_complete();
 
-#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
 	OPENSSL_config(NULL);
 #else
 	OPENSSL_init_crypto(OPENSSL_INIT_ADD_ALL_CIPHERS |
