diff --git a/include/pkcs11-helper-1.0/pkcs11h-openssl.h b/include/pkcs11-helper-1.0/pkcs11h-openssl.h
index ca1048b..c554f2a 100644
--- a/include/pkcs11-helper-1.0/pkcs11h-openssl.h
+++ b/include/pkcs11-helper-1.0/pkcs11h-openssl.h
@@ -67,6 +67,9 @@
 #ifndef __PKCS11H_HELPER_H
 #define __PKCS11H_HELPER_H
 
+#ifndef OPENSSL_NO_RSA
+#include <openssl/rsa.h>
+#endif
 #include <openssl/x509.h>
 #include <pkcs11-helper-1.0/pkcs11h-core.h>
 #include <pkcs11-helper-1.0/pkcs11h-certificate.h>
diff --git a/lib/_pkcs11h-crypto-openssl.c b/lib/_pkcs11h-crypto-openssl.c
index ce073aa..c94a188 100644
--- a/lib/_pkcs11h-crypto-openssl.c
+++ b/lib/_pkcs11h-crypto-openssl.c
@@ -73,6 +73,11 @@
 #endif
 #endif
 
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+#define X509_get_notAfter X509_get0_notAfter
+#define X509_get_notBefore X509_get0_notBefore
+#endif
+
 #if OPENSSL_VERSION_NUMBER < 0x00908000L
 typedef unsigned char *__pkcs11_openssl_d2i_t;
 #else
@@ -86,7 +91,9 @@ __pkcs11h_crypto_openssl_initialize (
 ) {
 	(void)global_data;
 
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
 	OpenSSL_add_all_digests ();
+#endif
 
 	return TRUE;
 }
@@ -111,8 +118,8 @@ __pkcs11h_crypto_openssl_certificate_get_expiration (
 ) {
 	X509 *x509 = NULL;
 	__pkcs11_openssl_d2i_t d2i;
-	ASN1_TIME *notBefore;
-	ASN1_TIME *notAfter;
+	const ASN1_TIME *notBefore;
+	const ASN1_TIME *notAfter;
 
 	(void)global_data;
 
diff --git a/lib/pkcs11h-openssl.c b/lib/pkcs11h-openssl.c
index 3977b0a..6b53729 100644
--- a/lib/pkcs11h-openssl.c
+++ b/lib/pkcs11h-openssl.c
@@ -54,6 +54,9 @@
 #if defined(ENABLE_PKCS11H_OPENSSL)
 
 #include <pkcs11-helper-1.0/pkcs11h-openssl.h>
+#ifndef OPENSSL_NO_DSA
+#include <openssl/dsa.h>
+#endif
 #include "_pkcs11h-core.h"
 #include "_pkcs11h-mem.h"
 
@@ -1119,7 +1122,9 @@ pkcs11h_openssl_createSession (
 		"PKCS#11: pkcs11h_openssl_createSession - entry"
 	);
 
+#if OPENSSL_VERSION_NUMBER <0x10100000L
 	OpenSSL_add_all_digests ();
+#endif
 
 	if (
 		_pkcs11h_mem_malloc (
