diff --git a/mutt_ssl.c b/mutt_ssl.c
index eacd155..0d2f162 100644
--- a/mutt_ssl.c
+++ b/mutt_ssl.c
@@ -28,6 +28,12 @@
 #include <openssl/rand.h>
 #include <openssl/evp.h>
 
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+#define X509_get_notBefore X509_get0_notBefore
+#define X509_get_notAfter X509_get0_notAfter
+#define X509_STORE_CTX_get_chain X509_STORE_CTX_get1_chain
+#endif
+
 #undef _
 
 #include <string.h>
@@ -331,8 +337,10 @@ static int ssl_init (void)
 
   /* I don't think you can do this just before reading the error. The call
    * itself might clobber the last SSL error. */
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
   SSL_load_error_strings();
   SSL_library_init();
+#endif
   init_complete = 1;
   return 0;
 }
@@ -740,7 +748,7 @@ static void x509_fingerprint (char *s, int l, X509 * cert, const EVP_MD *(*hashf
   }
 }
 
-static char *asn1time_to_string (ASN1_UTCTIME *tm)
+static char *asn1time_to_string (const ASN1_UTCTIME *tm)
 {
   static char buf[64];
   BIO *bio;
