diff --git a/Platforms/Ogre/OgrePlatform/src/MyGUI_OgreDataManager.cpp b/Platforms/Ogre/OgrePlatform/src/MyGUI_OgreDataManager.cpp
index 0b05cdf..467df78 100644
--- a/Platforms/Ogre/OgrePlatform/src/MyGUI_OgreDataManager.cpp
+++ b/Platforms/Ogre/OgrePlatform/src/MyGUI_OgreDataManager.cpp
@@ -106,7 +106,7 @@ namespace MyGUI
 			if (!pFileInfo->empty())
 				pFileInfos.push_back(pFileInfo);
 			else
-				pFileInfo.setNull();
+				pFileInfo.reset();
 		}
 
 		result.reserve(resultSize);
@@ -134,7 +134,7 @@ namespace MyGUI
 				}
 			}
 
-			pFileInfo.setNull();
+			pFileInfo.reset();
 		}
 
 		return result;
diff --git a/Platforms/Ogre/OgrePlatform/src/MyGUI_OgreDataStream.cpp b/Platforms/Ogre/OgrePlatform/src/MyGUI_OgreDataStream.cpp
index 82ec1ac..61b001b 100644
--- a/Platforms/Ogre/OgrePlatform/src/MyGUI_OgreDataStream.cpp
+++ b/Platforms/Ogre/OgrePlatform/src/MyGUI_OgreDataStream.cpp
@@ -16,22 +16,22 @@ namespace MyGUI
 
 	OgreDataStream::~OgreDataStream()
 	{
-		mStream.setNull();
+		mStream.reset();
 	}
 
 	bool OgreDataStream::eof()
 	{
-		return mStream.isNull() ? true : mStream->eof();
+		return mStream ? true : mStream->eof();
 	}
 
 	size_t OgreDataStream::size()
 	{
-		return mStream.isNull() ? 0 : mStream->size();
+		return mStream ? 0 : mStream->size();
 	}
 
 	void OgreDataStream::readline(std::string& _source, Char _delim)
 	{
-		if (mStream.isNull())
+		if (mStream)
 		{
 			_source.clear();
 			return;
@@ -41,7 +41,7 @@ namespace MyGUI
 
 	size_t OgreDataStream::read(void* _buf, size_t _count)
 	{
-		if (mStream.isNull()) return 0;
+		if (mStream) return 0;
 		return mStream->read(_buf, _count);
 	}
 
diff --git a/Platforms/Ogre/OgrePlatform/src/MyGUI_OgreRenderManager.cpp b/Platforms/Ogre/OgrePlatform/src/MyGUI_OgreRenderManager.cpp
index 2948ab4..07ec897 100644
--- a/Platforms/Ogre/OgrePlatform/src/MyGUI_OgreRenderManager.cpp
+++ b/Platforms/Ogre/OgrePlatform/src/MyGUI_OgreRenderManager.cpp
@@ -293,7 +293,7 @@ namespace MyGUI
 		{
 			OgreTexture* texture = static_cast<OgreTexture*>(_texture);
 			Ogre::TexturePtr texture_ptr = texture->getOgreTexture();
-			if (!texture_ptr.isNull())
+			if (!texture_ptr)
 			{
 				mRenderSystem->_setTexture(0, true, texture_ptr);
 				mRenderSystem->_setTextureUnitFiltering(0, Ogre::FO_LINEAR, Ogre::FO_LINEAR, Ogre::FO_NONE);
@@ -387,7 +387,7 @@ namespace MyGUI
 		if (item == mTextures.end())
 		{
 			Ogre::TexturePtr texture = (Ogre::TexturePtr)Ogre::TextureManager::getSingleton().getByName(_name);
-			if (!texture.isNull())
+			if (!texture)
 			{
 				ITexture* result = createTexture(_name);
 				static_cast<OgreTexture*>(result)->setOgreTexture(texture);
diff --git a/Platforms/Ogre/OgrePlatform/src/MyGUI_OgreTexture.cpp b/Platforms/Ogre/OgrePlatform/src/MyGUI_OgreTexture.cpp
index 45ec397..c487233 100644
--- a/Platforms/Ogre/OgrePlatform/src/MyGUI_OgreTexture.cpp
+++ b/Platforms/Ogre/OgrePlatform/src/MyGUI_OgreTexture.cpp
@@ -69,10 +69,10 @@ namespace MyGUI
 			mRenderTarget = nullptr;
 		}
 
-		if (!mTexture.isNull())
+		if (!mTexture)
 		{
 			Ogre::TextureManager::getSingleton().remove(mTexture->getName());
-			mTexture.setNull();
+			mTexture.reset();
 		}
 	}
 
@@ -254,7 +254,7 @@ namespace MyGUI
 		mPixelFormat = Ogre::PF_UNKNOWN;
 		mNumElemBytes = 0;
 
-		if (!mTexture.isNull())
+		if (!mTexture)
 		{
 			mPixelFormat = mTexture->getFormat();
 
diff --git a/Platforms/Ogre/OgrePlatform/src/MyGUI_OgreVertexBuffer.cpp b/Platforms/Ogre/OgrePlatform/src/MyGUI_OgreVertexBuffer.cpp
index 9079176..6ea50ec 100644
--- a/Platforms/Ogre/OgrePlatform/src/MyGUI_OgreVertexBuffer.cpp
+++ b/Platforms/Ogre/OgrePlatform/src/MyGUI_OgreVertexBuffer.cpp
@@ -57,7 +57,7 @@ namespace MyGUI
 	{
 		delete mRenderOperation.vertexData;
 		mRenderOperation.vertexData = 0;
-		mVertexBuffer.setNull();
+		mVertexBuffer.reset();
 	}
 
 	void OgreVertexBuffer::resizeVertexBuffer()
