diff --git a/src/_cffi_src/openssl/asn1.py b/src/_cffi_src/openssl/asn1.py
index 51acf10..9f63574 100644
--- a/src/_cffi_src/openssl/asn1.py
+++ b/src/_cffi_src/openssl/asn1.py
@@ -100,4 +100,7 @@ ASN1_TYPE *d2i_ASN1_TYPE(ASN1_TYPE **, const unsigned char **, long);
 """
 
 CUSTOMIZATIONS = """
+#if OPENSSL_API_COMPAT >= 0x10100000L
+#define ASN1_STRING_data ASN1_STRING_get0_data
+#endif
 """
diff --git a/src/_cffi_src/openssl/bignum.py b/src/_cffi_src/openssl/bignum.py
index 2c140c9..487b768 100644
--- a/src/_cffi_src/openssl/bignum.py
+++ b/src/_cffi_src/openssl/bignum.py
@@ -87,4 +87,8 @@ int BN_rshift1(BIGNUM *, BIGNUM *);
 """
 
 CUSTOMIZATIONS = """
+#if OPENSSL_API_COMPAT >= 0x10100000L
+#undef BN_zero
+#define BN_zero(x) (0)
+#endif
 """
diff --git a/src/_cffi_src/openssl/callbacks.py b/src/_cffi_src/openssl/callbacks.py
index cf2b2da..834d9ba 100644
--- a/src/_cffi_src/openssl/callbacks.py
+++ b/src/_cffi_src/openssl/callbacks.py
@@ -118,6 +118,7 @@ static void init_mutexes(void) {
 
 
 int _setup_ssl_threads(void) {
+#if OPENSSL_API_COMPAT < 0x10100000L
     if (_ssl_locks == NULL) {
         _ssl_locks_count = CRYPTO_num_locks();
         _ssl_locks = calloc(_ssl_locks_count, sizeof(Cryptography_mutex));
@@ -130,6 +131,7 @@ int _setup_ssl_threads(void) {
         pthread_atfork(NULL, NULL, &init_mutexes);
 #endif
     }
+#endif
     return 1;
 }
 
diff --git a/src/_cffi_src/openssl/conf.py b/src/_cffi_src/openssl/conf.py
index 9db0162..4eae213 100644
--- a/src/_cffi_src/openssl/conf.py
+++ b/src/_cffi_src/openssl/conf.py
@@ -18,4 +18,9 @@ void OPENSSL_no_config(void);
 """
 
 CUSTOMIZATIONS = """
+#if OPENSSL_API_COMPAT >= 0x10100000L
+#define OPENSSL_config(n) CONF_modules_load(NULL,n,0)
+#define OPENSSL_no_config() \
+    OPENSSL_init_crypto(OPENSSL_INIT_NO_LOAD_CONFIG, NULL)
+#endif
 """
diff --git a/src/_cffi_src/openssl/crypto.py b/src/_cffi_src/openssl/crypto.py
index dfff21b..2fb28b2 100644
--- a/src/_cffi_src/openssl/crypto.py
+++ b/src/_cffi_src/openssl/crypto.py
@@ -133,6 +133,16 @@ int Cryptography_CRYPTO_set_mem_functions(
 }
 #endif
 
+#if OPENSSL_API_COMPAT >= 0x10100000L
+#define ASN1_STRING_data ASN1_STRING_get0_data
+#define CRYPTO_cleanup_all_ex_data() while(0) continue
+#define CRYPTO_num_locks()            (1)
+#define CRYPTO_set_locking_callback(func)
+#define CRYPTO_get_locking_callback()         (NULL)
+#define CRYPTO_set_add_lock_callback(func)
+#define CRYPTO_get_add_lock_callback()        (NULL)
+#endif
+
 void *Cryptography_malloc_wrapper(size_t size, const char *path, int line) {
     return malloc(size);
 }
diff --git a/src/_cffi_src/openssl/engine.py b/src/_cffi_src/openssl/engine.py
index 45ce952..de1a549 100644
--- a/src/_cffi_src/openssl/engine.py
+++ b/src/_cffi_src/openssl/engine.py
@@ -130,4 +130,30 @@ void ENGINE_cleanup(void);
 """
 
 CUSTOMIZATIONS = """
+#if OPENSSL_API_COMPAT >= 0x10100000L
+/*
+ * This function previously cleaned up anything that needs it. Auto-deinit will
+ * now take care of it so it is no longer required to call this function.
+ */
+# define ENGINE_cleanup() while(0) continue
+#endif
+
+#if OPENSSL_API_COMPAT >= 0x10100000L
+#define ENGINE_load_openssl() \
+    OPENSSL_init_crypto(OPENSSL_INIT_ENGINE_OPENSSL, NULL)
+#define ENGINE_load_dynamic() \
+    OPENSSL_init_crypto(OPENSSL_INIT_ENGINE_DYNAMIC, NULL)
+#ifndef OPENSSL_NO_STATIC_ENGINE
+#define ENGINE_load_padlock() \
+    OPENSSL_init_crypto(OPENSSL_INIT_ENGINE_PADLOCK, NULL)
+#define ENGINE_load_capi() \
+    OPENSSL_init_crypto(OPENSSL_INIT_ENGINE_CAPI, NULL)
+#define ENGINE_load_afalg() \
+    OPENSSL_init_crypto(OPENSSL_INIT_ENGINE_AFALG, NULL)
+#endif
+#define ENGINE_load_cryptodev() \
+    OPENSSL_init_crypto(OPENSSL_INIT_ENGINE_CRYPTODEV, NULL)
+#define ENGINE_load_rdrand() \
+    OPENSSL_init_crypto(OPENSSL_INIT_ENGINE_RDRAND, NULL)
+#endif
 """
diff --git a/src/_cffi_src/openssl/err.py b/src/_cffi_src/openssl/err.py
index 5770429..fe6fe4f 100644
--- a/src/_cffi_src/openssl/err.py
+++ b/src/_cffi_src/openssl/err.py
@@ -273,4 +273,10 @@ static const long Cryptography_HAS_RSA_R_PKCS_DECODING_ERROR = 1;
 static const long Cryptography_HAS_RSA_R_PKCS_DECODING_ERROR = 0;
 static const long RSA_R_PKCS_DECODING_ERROR = 0;
 #endif
+
+#if OPENSSL_API_COMPAT >= 0x10100000L
+#define ERR_load_crypto_strings() \
+    OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CRYPTO_STRINGS, NULL)
+#define ERR_free_strings() while(0) continue
+#endif
 """
diff --git a/src/_cffi_src/openssl/evp.py b/src/_cffi_src/openssl/evp.py
index 715bfe1..7bc7777 100644
--- a/src/_cffi_src/openssl/evp.py
+++ b/src/_cffi_src/openssl/evp.py
@@ -250,4 +250,37 @@ int (*EVP_PKEY_set1_tls_encodedpoint)(EVP_PKEY *, const unsigned char *,
 #ifndef EVP_PKEY_X25519
 #define EVP_PKEY_X25519 NID_X25519
 #endif
+
+#if OPENSSL_API_COMPAT >= 0x10100000L
+#define EVP_CIPHER_CTX_init(c)      EVP_CIPHER_CTX_reset(c)
+#define EVP_CIPHER_CTX_cleanup(c)   EVP_CIPHER_CTX_reset(c)
+#endif
+
+#if OPENSSL_API_COMPAT >= 0x10100000L
+#define OPENSSL_add_all_algorithms_conf() \
+    OPENSSL_init_crypto(OPENSSL_INIT_ADD_ALL_CIPHERS \
+                        | OPENSSL_INIT_ADD_ALL_DIGESTS \
+                        | OPENSSL_INIT_LOAD_CONFIG, NULL)
+#define OPENSSL_add_all_algorithms_noconf() \
+    OPENSSL_init_crypto(OPENSSL_INIT_ADD_ALL_CIPHERS \
+                        | OPENSSL_INIT_ADD_ALL_DIGESTS, NULL)
+
+#ifdef OPENSSL_LOAD_CONF
+#define OpenSSL_add_all_algorithms() \
+    OPENSSL_init_crypto(OPENSSL_INIT_ADD_ALL_CIPHERS \
+                        | OPENSSL_INIT_ADD_ALL_DIGESTS \
+                        | OPENSSL_INIT_LOAD_CONFIG, NULL)
+#else
+#define OpenSSL_add_all_algorithms() \
+    OPENSSL_init_crypto(OPENSSL_INIT_ADD_ALL_CIPHERS \
+                        | OPENSSL_INIT_ADD_ALL_DIGESTS, NULL)
+#endif
+
+#define OpenSSL_add_all_ciphers() \
+    OPENSSL_init_crypto(OPENSSL_INIT_ADD_ALL_CIPHERS, NULL)
+#define OpenSSL_add_all_digests() \
+    OPENSSL_init_crypto(OPENSSL_INIT_ADD_ALL_DIGESTS, NULL)
+
+#define EVP_cleanup() while(0) continue
+#endif
 """
diff --git a/src/_cffi_src/openssl/objects.py b/src/_cffi_src/openssl/objects.py
index 265ac75..7e9c82a 100644
--- a/src/_cffi_src/openssl/objects.py
+++ b/src/_cffi_src/openssl/objects.py
@@ -38,4 +38,7 @@ void OBJ_cleanup(void);
 """
 
 CUSTOMIZATIONS = """
+#if OPENSSL_API_COMPAT >= 0x10100000L
+#define OBJ_cleanup() while(0) continue
+#endif
 """
diff --git a/src/_cffi_src/openssl/rand.py b/src/_cffi_src/openssl/rand.py
index 089da8b..b635d1c 100644
--- a/src/_cffi_src/openssl/rand.py
+++ b/src/_cffi_src/openssl/rand.py
@@ -32,4 +32,8 @@ void RAND_cleanup(void);
 
 CUSTOMIZATIONS = """
 static const long Cryptography_HAS_EGD = 0;
+
+#if OPENSSL_API_COMPAT >= 0x10100000L
+#define RAND_cleanup() while(0) continue
+#endif
 """
diff --git a/src/_cffi_src/openssl/ssl.py b/src/_cffi_src/openssl/ssl.py
index 2933114..91fb9f9 100644
--- a/src/_cffi_src/openssl/ssl.py
+++ b/src/_cffi_src/openssl/ssl.py
@@ -708,4 +708,151 @@ void (*SSL_CTX_set_psk_client_callback)(SSL_CTX *,
 #else
 static const long Cryptography_HAS_PSK = 1;
 #endif
+
+#if OPENSSL_API_COMPAT >= 0x10100000L
+#define SSL_library_init() OPENSSL_init_ssl(0, NULL)
+#define SSL_load_error_strings() \
+    OPENSSL_init_ssl(OPENSSL_INIT_LOAD_SSL_STRINGS \
+                     | OPENSSL_INIT_LOAD_CRYPTO_STRINGS, NULL)
+
+static void _set_protocol_version(const SSL_METHOD *m, int version)
+{
+    SSL_CTX *ctx = SSL_CTX_new(m);
+    SSL_CTX_set_min_proto_version(ctx, version);
+    SSL_CTX_set_max_proto_version(ctx, version);
+    SSL_CTX_free(ctx);
+}
+
+/*
+ * TLSv1_1 and TLSv1_2 are recent additions.  Only sufficiently new versions of
+ * OpenSSL support them.
+ */
+const SSL_METHOD *Cryptography_TLSv1_1_method(void)
+{
+    const SSL_METHOD *m = TLS_method();
+    _set_protocol_version(m, TLS1_1_VERSION);
+    return m;
+}
+
+const SSL_METHOD *Cryptography_TLSv1_1_server_method(void)
+{
+    const SSL_METHOD *m = TLS_server_method();
+    _set_protocol_version(m, TLS1_1_VERSION);
+    return m;
+}
+
+const SSL_METHOD *Cryptography_TLSv1_1_client_method(void)
+{
+    const SSL_METHOD *m = TLS_client_method();
+    _set_protocol_version(m, TLS1_1_VERSION);
+    return m;
+}
+
+const SSL_METHOD *Cryptography_TLSv1_2_method(void)
+{
+    const SSL_METHOD *m = TLS_method();
+    _set_protocol_version(m, TLS1_2_VERSION);
+    return m;
+}
+
+const SSL_METHOD *Cryptography_TLSv1_2_server_method(void)
+{
+    const SSL_METHOD *m = TLS_server_method();
+    _set_protocol_version(m, TLS1_2_VERSION);
+    return m;
+}
+
+const SSL_METHOD *Cryptography_TLSv1_2_client_method(void)
+{
+    const SSL_METHOD *m = TLS_client_method();
+    _set_protocol_version(m, TLS1_2_VERSION);
+    return m;
+}
+
+#define TLSv1_1_method Cryptography_TLSv1_1_method
+#define TLSv1_1_server_method Cryptography_TLSv1_1_server_method
+#define TLSv1_1_client_method Cryptography_TLSv1_1_client_method
+#define TLSv1_2_method Cryptography_TLSv1_2_method
+#define TLSv1_2_server_method Cryptography_TLSv1_2_server_method
+#define TLSv1_2_client_method Cryptography_TLSv1_2_client_method
+
+#ifndef OPENSSL_NO_SSL3_METHOD
+const SSL_METHOD *Cryptography_SSLv3_method(void)
+{
+    const SSL_METHOD *m = TLS_method();
+    _set_protocol_version(m, SSL3_VERSION);
+    return m;
+}
+
+const SSL_METHOD *Cryptography_SSLv3_server_method(void)
+{
+    const SSL_METHOD *m = TLS_server_method();
+    _set_protocol_version(m, SSL3_VERSION);
+    return m;
+}
+
+const SSL_METHOD *Cryptography_SSLv3_client_method(void)
+{
+    const SSL_METHOD *m = TLS_client_method();
+    _set_protocol_version(m, SSL3_VERSION);
+    return m;
+}
+
+#define SSLv3_method Cryptography_SSLv3_method
+#define SSLv3_server_method Cryptography_SSLv3_server_method
+#define SSLv3_client_method Cryptography_SSLv3_client_method
+#endif
+
+const SSL_METHOD *Cryptography_TLSv1_method(void)
+{
+    const SSL_METHOD *m = TLS_method();
+    _set_protocol_version(m, TLS1_VERSION);
+    return m;
+}
+
+const SSL_METHOD *Cryptography_TLSv1_server_method(void)
+{
+    const SSL_METHOD *m = TLS_server_method();
+    _set_protocol_version(m, TLS1_VERSION);
+    return m;
+}
+
+const SSL_METHOD *Cryptography_TLSv1_client_method(void)
+{
+    const SSL_METHOD *m = TLS_client_method();
+    _set_protocol_version(m, TLS1_VERSION);
+    return m;
+}
+
+#define TLSv1_method Cryptography_TLSv1_method
+#define TLSv1_server_method Cryptography_TLSv1_server_method
+#define TLSv1_client_method Cryptography_TLSv1_client_method
+
+#if !defined(OPENSSL_NO_DTLS) && !CRYPTOGRAPHY_OPENSSL_LESS_THAN_102
+const SSL_METHOD *Cryptography_DTLSv1_method(void)
+{
+    const SSL_METHOD *m = TLS_method();
+    _set_protocol_version(m, DTLS1_VERSION);
+    return m;
+}
+
+const SSL_METHOD *Cryptography_DTLSv1_server_method(void)
+{
+    const SSL_METHOD *m = TLS_server_method();
+    _set_protocol_version(m, DTLS1_VERSION);
+    return m;
+}
+
+const SSL_METHOD *Cryptography_DTLSv1_client_method(void)
+{
+    const SSL_METHOD *m = TLS_client_method();
+    _set_protocol_version(m, DTLS1_VERSION);
+    return m;
+}
+
+#define DTLSv1_method Cryptography_DTLSv1_method
+#define DTLSv1_server_method Cryptography_DTLSv1_server_method
+#define DTLSv1_client_method Cryptography_DTLSv1_client_method
+#endif
+#endif
 """
diff --git a/src/_cffi_src/openssl/x509.py b/src/_cffi_src/openssl/x509.py
index f4703c4..5c2e61f 100644
--- a/src/_cffi_src/openssl/x509.py
+++ b/src/_cffi_src/openssl/x509.py
@@ -433,4 +433,15 @@ const ASN1_INTEGER *X509_REVOKED_get0_serialNumber(const X509_REVOKED *x)
     return x->serialNumber;
 }
 #endif
+
+#if OPENSSL_API_COMPAT >= 0x10100000L
+#define X509_get_notBefore X509_getm_notBefore
+#define X509_get_notAfter X509_getm_notAfter
+#define X509_set_notBefore X509_set1_notBefore
+#define X509_set_notAfter X509_set1_notAfter
+#define X509_CRL_get_lastUpdate X509_CRL_get0_lastUpdate
+#define X509_CRL_get_nextUpdate X509_CRL_get0_nextUpdate
+#define X509_CRL_set_lastUpdate X509_CRL_set1_lastUpdate
+#define X509_CRL_set_nextUpdate X509_CRL_set1_nextUpdate
+#endif
 """
diff --git a/src/_cffi_src/openssl/x509_vfy.py b/src/_cffi_src/openssl/x509_vfy.py
index 2971f2c..0f66f34 100644
--- a/src/_cffi_src/openssl/x509_vfy.py
+++ b/src/_cffi_src/openssl/x509_vfy.py
@@ -310,4 +310,11 @@ void (*X509_STORE_set_get_issuer)(X509_STORE *,
 #else
 static const long Cryptography_HAS_X509_STORE_CTX_GET_ISSUER = 1;
 #endif
+
+#if OPENSSL_API_COMPAT >= 0x10100000L
+#define X509_V_FLAG_CB_ISSUER_CHECK             0x0   /* Deprecated */
+#define X509_STORE_CTX_get_chain X509_STORE_CTX_get0_chain
+#define X509_STORE_CTX_set_chain X509_STORE_CTX_set0_untrusted
+#define X509_STORE_CTX_trusted_stack X509_STORE_CTX_set0_trusted_stack
+#endif
 """
