diff --git a/src/threadsupport.c b/src/threadsupport.c
index 775ae61..e328d7c 100644
--- a/src/threadsupport.c
+++ b/src/threadsupport.c
@@ -88,6 +88,7 @@ pycurl_release_thread(PyThreadState *state)
 
 #ifdef PYCURL_NEED_OPENSSL_TSL
 
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
 static PyThread_type_lock *pycurl_openssl_tsl = NULL;
 
 static void
@@ -105,10 +106,12 @@ pycurl_ssl_id(void)
 {
     return (unsigned long) PyThread_get_thread_ident();
 }
+#endif
 
 PYCURL_INTERNAL int
 pycurl_ssl_init(void)
 {
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
     int i, c = CRYPTO_num_locks();
 
     pycurl_openssl_tsl = PyMem_New(PyThread_type_lock, c);
@@ -132,12 +135,14 @@ pycurl_ssl_init(void)
 
     CRYPTO_set_id_callback(pycurl_ssl_id);
     CRYPTO_set_locking_callback(pycurl_ssl_lock);
+#endif
     return 0;
 }
 
 PYCURL_INTERNAL void
 pycurl_ssl_cleanup(void)
 {
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
     if (pycurl_openssl_tsl) {
         int i, c = CRYPTO_num_locks();
 
@@ -151,6 +156,7 @@ pycurl_ssl_cleanup(void)
         PyMem_Free(pycurl_openssl_tsl);
         pycurl_openssl_tsl = NULL;
     }
+#endif
 }
 #endif
 
