diff -ur a/visual_regexp.tcl b/visual_regexp.tcl
--- a/visual_regexp.tcl	2013-02-12 14:36:07.571112314 +0100
+++ b/visual_regexp.tcl	2013-10-29 10:00:18.519775484 +0100
@@ -496,7 +496,7 @@
 	if {$tcl_platform(platform) == "windows"} {
 		text .help.t -bd 2 -relief groove -font {Courier 10}
 	} else {
-		text .help.t -bd 2 -relief groove
+		text .help.t -bd 2 -relief groove -font {9x15}
 	}
 	pack .help.t -side top -padx 20
 	.help.t tag configure bold -font "[.help.t cget -font] bold"
