diff --git a/extract-cert.c b/extract-cert.c
index b071bf4..2213f28 100644
--- a/extract-cert.c
+++ b/extract-cert.c
@@ -91,8 +91,12 @@ int main(int argc, char **argv)
 {
 	char *cert_src;
 
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
 	OpenSSL_add_all_algorithms();
 	ERR_load_crypto_strings();
+#else
+	OPENSSL_init_crypto(OPENSSL_INIT_ADD_ALL_CIPHERS | OPENSSL_INIT_ADD_ALL_DIGESTS | OPENSSL_INIT_LOAD_CRYPTO_STRINGS, NULL);
+#endif
 	ERR_clear_error();
 
 	kbuild_verbose = atoi(getenv("KBUILD_VERBOSE")?:"0");
diff --git a/sign-file.c b/sign-file.c
index 53af6dc..50aa165 100755
--- a/sign-file.c
+++ b/sign-file.c
@@ -227,8 +227,12 @@ int main(int argc, char **argv)
 	X509 *x509;
 	BIO *bd, *bm;
 	int opt, n;
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
 	OpenSSL_add_all_algorithms();
 	ERR_load_crypto_strings();
+#else
+	OPENSSL_init_crypto(OPENSSL_INIT_ADD_ALL_CIPHERS | OPENSSL_INIT_ADD_ALL_DIGESTS | OPENSSL_INIT_LOAD_CRYPTO_STRINGS, NULL);
+#endif
 	ERR_clear_error();
 
 	key_pass = getenv("KBUILD_SIGN_PIN");
@@ -296,7 +300,11 @@ int main(int argc, char **argv)
 		x509 = read_x509(x509_name);
 
 		/* Digest the module data. */
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
 		OpenSSL_add_all_digests();
+#else
+		OPENSSL_init_crypto(OPENSSL_INIT_ADD_ALL_DIGESTS, NULL);
+#endif
 		display_openssl_errors(__LINE__);
 		digest_algo = EVP_get_digestbyname(hash_algo);
 		ERR(!digest_algo, "EVP_get_digestbyname");
