Index: libe2dbg/user/entry.c
===================================================================
--- libe2dbg/user/entry.c	(revision 1465)
+++ libe2dbg/user/entry.c	(working copy)
@@ -134,7 +134,7 @@
 			  void (*init) (void),
 			  void (*fini) (void),
 			  void (*rtld_fini) (void), 
-			  void *__unbounded stack_end)
+			  void *__attribute__ ((unbounded)) stack_end)
 {
   eresi_Addr		orig;
   int			(*libcstartmain)();
