diff -Naur eresi-0.82_beta2.orig//configure eresi-0.82_beta2/configure
--- eresi-0.82_beta2.orig//configure	2011-05-24 11:19:49.000000000 -0400
+++ eresi-0.82_beta2/configure	2011-05-24 11:31:31.000000000 -0400
@@ -779,29 +779,29 @@
 fi
 
 # Create links for 64bits mode
-echo "install64 :"						       >> Makefile
-echo "	@ln -sf \$(BINPATH)elfsh64         \$(BINPATH)elfsh"	       >> Makefile
-echo "	@ln -sf \$(BINPATH)e2dbg64         \$(BINPATH)e2dbg"	       >> Makefile
-echo "	@ln -sf \$(BINPATH)etrace64        \$(BINPATH)etrace"	       >> Makefile
-echo "	@ln -sf \$(BINPATH)evarista64      \$(BINPATH)evarista"	       >> Makefile
-echo "	@ln -sf \$(LIBPATH)libaspect64.a   \$(LIBPATH)libaspect.a"    >> Makefile
-echo "	@ln -sf \$(LIBPATH)libelfsh64.a    \$(LIBPATH)libelfsh.a"     >> Makefile
-echo "	@ln -sf \$(LIBPATH)libe2dbg64.a    \$(LIBPATH)libe2dbg.a"     >> Makefile
-echo "	@ln -sf \$(LIBPATH)libmjollnir64.a \$(LIBPATH)libmjollnir.a"  >> Makefile
-echo "	@ln -sf \$(LIBPATH)librevm64.a     \$(LIBPATH)librevm.a"      >> Makefile
-echo "	@ln -sf \$(LIBPATH)libstderesi64.a \$(LIBPATH)libstderesi.a"  >> Makefile
-echo "	@ln -sf \$(LIBPATH)libedfmt64.a    \$(LIBPATH)libedfmt.a"     >> Makefile
-echo "	@ln -sf \$(LIBPATH)libui64.a       \$(LIBPATH)libui.a"        >> Makefile
-
-echo "	@ln -sf \$(LIBPATH)libaspect64.so   \$(LIBPATH)libaspect.so"   >> Makefile
-echo "	@ln -sf \$(LIBPATH)libelfsh64.so    \$(LIBPATH)libelfsh.so"    >> Makefile
-echo "	@ln -sf \$(LIBPATH)libetrace64.so    \$(LIBPATH)libetrace.so"  >> Makefile
-echo "	@ln -sf \$(LIBPATH)libe2dbg64.so    \$(LIBPATH)libe2dbg.so"    >> Makefile
-echo "	@ln -sf \$(LIBPATH)libmjollnir64.so \$(LIBPATH)libmjollnir.so" >> Makefile
-echo "	@ln -sf \$(LIBPATH)librevm64.so     \$(LIBPATH)librevm.so"     >> Makefile
-echo "	@ln -sf \$(LIBPATH)libstderesi64.so \$(LIBPATH)libstderesi.so" >> Makefile
-echo "	@ln -sf \$(LIBPATH)libedfmt64.so    \$(LIBPATH)libedfmt.so"    >> Makefile
-echo "	@ln -sf \$(LIBPATH)libui64.so       \$(LIBPATH)libui.so"       >> Makefile
+echo "install64 :"							>> Makefile
+echo "	@cd \$(BINPATH) && ln -sf elfsh64 elfsh"			>> Makefile
+echo "	@cd \$(BINPATH) && ln -sf e2dbg64 e2dbg"			>> Makefile
+echo "	@cd \$(BINPATH) && ln -sf etrace64 etrace"			>> Makefile
+echo "	@cd \$(BINPATH) && ln -sf evarista64 evarista"			>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libaspect64.a   libaspect.a"		>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libelfsh64.a    libelfsh.a"		>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libe2dbg64.a    libe2dbg.a"		>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libmjollnir64.a libmjollnir.a"	>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf librevm64.a     librevm.a"		>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libstderesi64.a libstderesi.a"	>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libedfmt64.a    libedfmt.a"		>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libui64.a       libui.a"		>> Makefile
+
+echo "	@cd \$(LIBPATH) && ln -sf libaspect64.so   libaspect.so"	>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libelfsh64.so    libelfsh.so"		>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libetrace64.so   libetrace.so"	>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libe2dbg64.so    libe2dbg.so"		>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libmjollnir64.so libmjollnir.so"	>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf librevm64.so     librevm.so"		>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libstderesi64.so libstderesi.so"	>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libedfmt64.so    libedfmt.so"		>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libui64.so       libui.so"		>> Makefile
 
 echo "install : prepareDir #mod_install "			>> Makefile
 echo "	@cp elfsh/elfsh?? \$(BINPATH)/ "			>> Makefile
@@ -828,31 +828,31 @@
 
 
 # Create links
-echo "	@ln -sf \$(BINPATH)elfsh32         \$(BINPATH)elfsh"	      >> Makefile
-echo "	@ln -sf \$(BINPATH)e2dbg32         \$(BINPATH)e2dbg"	      >> Makefile
-echo "	@ln -sf \$(BINPATH)etrace32        \$(BINPATH)etrace"	      >> Makefile
-echo "	@ln -sf \$(BINPATH)evarista32      \$(BINPATH)evarista"	       >> Makefile
-echo "	@ln -sf \$(BINPATH)kernsh32        \$(BINPATH)kernsh"        >> Makefile
-echo "	@ln -sf \$(LIBPATH)libaspect32.a   \$(LIBPATH)libaspect.a"    >> Makefile
-echo "	@ln -sf \$(LIBPATH)libelfsh32.a    \$(LIBPATH)libelfsh.a"     >> Makefile
-echo "	@ln -sf \$(LIBPATH)libetrace32.a    \$(LIBPATH)libetrace.a"     >> Makefile
-echo "	@ln -sf \$(LIBPATH)libe2dbg32.a    \$(LIBPATH)libe2dbg.a"     >> Makefile
-echo "	@ln -sf \$(LIBPATH)libmjollnir32.a \$(LIBPATH)libmjollnir.a"  >> Makefile
-echo "	@ln -sf \$(LIBPATH)librevm32.a     \$(LIBPATH)librevm.a"      >> Makefile
-echo "	@ln -sf \$(LIBPATH)libstderesi32.a \$(LIBPATH)libstderesi.a"      >> Makefile
-echo "	@ln -sf \$(LIBPATH)libedfmt32.a    \$(LIBPATH)libedfmt.a"     >> Makefile
-echo "	@ln -sf \$(LIBPATH)libui32.a       \$(LIBPATH)libui.a"        >> Makefile
-echo "	@ln -sf \$(LIBPATH)libaspect32.so   \$(LIBPATH)libaspect.so"   >> Makefile
-echo "	@ln -sf \$(LIBPATH)libelfsh32.so    \$(LIBPATH)libelfsh.so"    >> Makefile
-echo "	@ln -sf \$(LIBPATH)libetrace32.so    \$(LIBPATH)libetrace.so"    >> Makefile
-echo "	@ln -sf \$(LIBPATH)libe2dbg32.so    \$(LIBPATH)libe2dbg.so"    >> Makefile
-echo "	@ln -sf \$(LIBPATH)libmjollnir32.so \$(LIBPATH)libmjollnir.so" >> Makefile
-echo "	@ln -sf \$(LIBPATH)librevm32.so     \$(LIBPATH)librevm.so"     >> Makefile
-echo "	@ln -sf \$(LIBPATH)libstderesi32.so \$(LIBPATH)libstderesi.so" >> Makefile
-echo "	@ln -sf \$(LIBPATH)libedfmt32.so    \$(LIBPATH)libedfmt.so"    >> Makefile
-echo "	@ln -sf \$(LIBPATH)libui32.so       \$(LIBPATH)libui.so"       >> Makefile
-echo "	@ln -sf \$(LIBPATH)libkernsh32.a   \$(LIBPATH)libkernsh.a"   >> Makefile
-echo "	@ln -sf \$(LIBPATH)libkernsh32.so   \$(LIBPATH)libkernsh.so"   >> Makefile
+echo "	@cd \$(BINPATH) && ln -sf elfsh32         elfsh"		>> Makefile
+echo "	@cd \$(BINPATH) && ln -sf e2dbg32         e2dbg"		>> Makefile
+echo "	@cd \$(BINPATH) && ln -sf etrace32        etrace"		>> Makefile
+echo "	@cd \$(BINPATH) && ln -sf evarista32      evarista"		>> Makefile
+echo "	@cd \$(BINPATH) && ln -sf kernsh32        kernsh"		>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libaspect32.a   libaspect.a"		>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libelfsh32.a    libelfsh.a"		>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libetrace32.a   libetrace.a"		>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libe2dbg32.a    libe2dbg.a"		>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libmjollnir32.a libmjollnir.a"	>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf librevm32.a     librevm.a"		>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libstderesi32.a libstderesi.a"	>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libedfmt32.a    libedfmt.a"		>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libui32.a       libui.a"		>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libaspect32.so   libaspect.so"	>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libelfsh32.so    libelfsh.so"		>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libetrace32.so   libetrace.so"	>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libe2dbg32.so    libe2dbg.so"		>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libmjollnir32.so libmjollnir.so"	>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf librevm32.so     librevm.so"		>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libstderesi32.so libstderesi.so"	>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libedfmt32.so    libedfmt.so"		>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libui32.so       libui.so"		>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libkernsh32.a    libkernsh.a"		>> Makefile
+echo "	@cd \$(LIBPATH) && ln -sf libkernsh32.so   libkernsh.so"	>> Makefile
 
 # Install headers
 echo "	@cp -R libelfsh/include/* \$(INCPATH) "			>> Makefile
