--- splashutils-1.5.4.4/src/ttf.c  2014-12-21 06:43:31.242847229 +0100
+++ splashutils-1.5.4.4/src/ttf.c   2014-12-21 06:45:05.235647523 +0100
@@ -25,8 +25,8 @@
 #include <assert.h>
 
 #include <ft2build.h>
-#include <freetype/ftoutln.h>
-#include <freetype/ttnameid.h>
+#include FT_OUTLINE_H
+#include FT_TRUETYPE_IDS_H
 
 #include "common.h"
 #include "render.h"
--- splashutils-1.5.4.4/src/ttf.h  2014-12-21 06:43:37.474834083 +0100
+++ splashutils-1.5.4.4/src/ttf.h   2014-12-21 06:45:11.899633254 +0100
@@ -1,8 +1,8 @@
 #ifndef _TTF_H
 #define _TTF_H
 #include <ft2build.h>
-#include <freetype/ftoutln.h>
-#include <freetype/ttnameid.h>
+#include FT_OUTLINE_H
+#include FT_TRUETYPE_IDS_H
 
 #define CACHED_METRICS  0x10
 #define CACHED_BITMAP   0x01
