--- a/opencv-2.4.9/cmake/OpenCVDetectCUDA.cmake	2015-02-27 00:35:27.675722751 +0100
+++ b/opencv-2.4.9/cmake/OpenCVDetectCUDA.cmake	2015-02-27 00:35:23.402723122 +0100
@@ -96,6 +96,8 @@
     else()
       if(${CUDA_VERSION} VERSION_LESS "5.0")
         set(__cuda_arch_bin "1.1 1.2 1.3 2.0 2.1(2.0) 3.0")
+      elseif(${CUDA_VERSION} VERSION_GREATER "6.5")
+        set(__cuda_arch_bin "2.0 2.1(2.0) 3.0 3.5")
       else()
         set(__cuda_arch_bin "1.1 1.2 1.3 2.0 2.1(2.0) 3.0 3.5")
       endif()
