--- elinks-0.12pre6-orig/configure.in	2017-10-29 21:52:52.000000000 +1100
+++ elinks-0.12pre6/configure.in	2017-10-29 22:04:18.428702527 +1100
@@ -1026,6 +1026,9 @@
 			if test "$cf_result" != yes; then
 				AC_LINK_IFELSE([AC_LANG_PROGRAM([[#include <openssl/ssl.h>]], [[SSLeay_add_ssl_algorithms()]])],[cf_result=yes],[cf_result=no])
 			fi
+			if test "$cf_result" != yes; then
+				AC_LINK_IFELSE([AC_LANG_PROGRAM([[#include <openssl/crypto.h>]], [[OpenSSL_version_num()]])],[cf_result=yes],[cf_result=no])
+			fi
 		fi
 	done
 
@@ -1040,6 +1043,8 @@
 
 		CFLAGS="$CFLAGS_X"
 		AC_SUBST(OPENSSL_CFLAGS)
+
+		AC_CHECK_FUNCS([SSL_set_options] [SSL_get_SSL_CTX])
 	fi
 fi
 
--- elinks-0.12pre6-orig/src/network/ssl/ssl.c	2017-10-29 21:52:52.000000000 +1100
+++ elinks-0.12pre6/src/network/ssl/ssl.c	2017-10-29 22:04:18.429702527 +1100
@@ -7,6 +7,7 @@
 #ifdef CONFIG_OPENSSL
 #include <openssl/ssl.h>
 #include <openssl/rand.h>
+#include <openssl/opensslv.h>
 #elif defined(CONFIG_GNUTLS)
 #include <gnutls/gnutls.h>
 #include <gnutls/x509.h>
@@ -61,7 +62,9 @@
 	}
 #endif
 
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
 	SSLeay_add_ssl_algorithms();
+#endif
 	context = SSL_CTX_new(SSLv23_client_method());
 	SSL_CTX_set_options(context, SSL_OP_ALL);
 	SSL_CTX_set_default_verify_paths(context);
